used-before-def incorrectly flagged when using match/case with a walrus operator

**Bug Report**

When you use a walrus operator in a `match` statement mypy will complain about an used-before-def for any use of the variable after the first case.  (It doesn't complain for the first case)

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.11&gist=f0900514987ceef340db8076b6319cd8

**Expected Behavior**

I'm pretty sure the code is fine and that there's no way the variable is actually undefined.  I also realize that I can rewrite the code to be `case 5 as value:` to work around the issue.  Also, I realize the example is trivial but imagine if the `5` was `int()`.
    
**Actual Behavior**

<!-- What went wrong? Paste mypy's output. -->

main.py:6: error: Name "result" is used before definition  [used-before-def]

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used:  mypy 1.0.0 (compiled: yes)
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.11

--- a/mypy/partially_defined.py
+++ b/mypy/partially_defined.py
@@ -396,8 +396,8 @@ class PossiblyUndefinedVariableVisitor(ExtendedTraverserVisitor):
         self.tracker.end_branch_statement()
 
     def visit_match_stmt(self, o: MatchStmt) -> None:
-        self.tracker.start_branch_statement()
         o.subject.accept(self)
+        self.tracker.start_branch_statement()
         for i in range(len(o.patterns)):
             pattern = o.patterns[i]
             pattern.accept(self)

--- a/test-data/unit/check-python310.test
+++ b/test-data/unit/check-python310.test
@@ -1826,7 +1826,6 @@ def f0(x: int) -> None:
             c = y
     d = y
 
-
 [case testTypeAliasWithNewUnionSyntaxAndNoneLeftOperand]
 from typing import overload
 class C:

--- a/test-data/unit/check-python310.test
+++ b/test-data/unit/check-python310.test
@@ -1813,6 +1813,20 @@ def f1(x: int) -> int:
 
 [typing fixtures/typing-medium.pyi]
 
+[case testUsedBeforeDefMatchWalrus]
+# flags: --enable-error-code used-before-def
+import typing
+
+def f0(x: int) -> None:
+    a = y  # E: Cannot determine type of "y"  # E: Name "y" is used before definition
+    match y := x:
+        case 1:
+            b = y
+        case 2:
+            c = y
+    d = y
+
+
 [case testTypeAliasWithNewUnionSyntaxAndNoneLeftOperand]
 from typing import overload
 class C:

